From 23db373c78074315a082590879bd79006ce45232 Mon Sep 17 00:00:00 2001
From: Kevin O'Connor <kevin@koconnor.net>
Date: Mon, 28 Sep 2026 22:06:15 -0400
Subject: [PATCH 1/2] ar100: Utilize serial_irq.c code for message parsing and
 generation

Utilize the generic code.  The previous code could possibly busy wait
for a long time during console_sendf().

Signed-off-by: Kevin O'Connor <kevin@koconnor.net>
(cherry picked from commit 7675eeca128c46448c9eea480b804ebcfa670a13)
(cherry picked from commit 6fd02d169f3720543c56007b33d5be8e9955105b)
---
 src/ar100/Makefile |  2 +-
 src/ar100/main.c   | 66 ++++++++++++++--------------------------------
 src/ar100/serial.c | 12 +++------
 src/ar100/serial.h |  4 +--
 4 files changed, 26 insertions(+), 58 deletions(-)

diff --git a/src/ar100/Makefile b/src/ar100/Makefile
index 166e88c99..2024ab0b8 100644
--- a/src/ar100/Makefile
+++ b/src/ar100/Makefile
@@ -17,7 +17,7 @@ SFLAGS		+= -Ilib/ar100
 # Add source files
 src-y += ar100/main.c ar100/gpio.c ar100/serial.c
 src-y += ar100/util.c ar100/timer.c
-src-y += generic/crc16_ccitt.c generic/timer_irq.c
+src-y += generic/crc16_ccitt.c generic/timer_irq.c generic/serial_irq.c
 
 OBJS_klipper.elf += $(OUT)lib/ar100/start.o
 OBJS_klipper.elf += $(OUT)lib/ar100/runtime.o
diff --git a/src/ar100/main.c b/src/ar100/main.c
index d4ca8aca8..0e2693acc 100644
--- a/src/ar100/main.c
+++ b/src/ar100/main.c
@@ -9,6 +9,7 @@
 #include "board/misc.h" // dynmem_start
 #include "board/irq.h"  // irq_disable
 #include "command.h"    // shutdown
+#include "generic/serial_irq.h" // serial_enable_tx_irq
 #include "generic/timer_irq.h"  // timer_dispatch_many
 #include "sched.h"      // sched_main
 
@@ -39,11 +40,28 @@ dynmem_end(void)
     return &dynmem_pool[sizeof(dynmem_pool)];
 }
 
+static int need_serial_tx;
+
 static void
 check_serial_pending(void)
 {
-    if(r_uart_fifo_rcv())
-        sched_wake_task(&console_wake);
+    while (r_uart_fifo_rcv())
+        serial_rx_byte(r_uart_getc());
+    while (need_serial_tx && r_uart_fifo_cantx()) {
+        uint8_t b;
+        int ret = serial_get_tx_byte(&b);
+        if (ret < 0)
+            need_serial_tx = 0;
+        else
+            r_uart_putc(b);
+    }
+}
+
+void
+serial_enable_tx_irq(void)
+{
+    need_serial_tx = 1;
+    check_serial_pending();
 }
 
 void
@@ -95,50 +113,6 @@ irq_poll(void)
     check_serial_pending();
 }
 
-/****************************************************************
-* Console IO
-****************************************************************/
-
-// Process any incoming commands
-void
-console_task(void)
-{
-    if (!sched_check_wake(&console_wake))
-        return;
-
-    int ret = 0;
-    for(int i=0; i<r_uart_fifo_rcv(); i++) {
-        receive_buf[receive_pos + ret++] = r_uart_getc();
-    }
-    if(!ret)
-        return;
-
-    int len = receive_pos + ret;
-    uint_fast8_t pop_count, msglen = len > MESSAGE_MAX ? MESSAGE_MAX : len;
-    ret = command_find_and_dispatch(receive_buf, msglen, &pop_count);
-    if (ret) {
-        len -= pop_count;
-        if (len) {
-            memcpy(receive_buf, &receive_buf[pop_count], len);
-            sched_wake_task(&console_wake);
-        }
-    }
-    receive_pos = len;
-}
-DECL_TASK(console_task);
-
-// Encode and transmit a "response" message
-void
-console_sendf(const struct command_encoder *ce, va_list args)
-{
-    uint8_t buf[MESSAGE_MAX];
-    uint_fast8_t msglen = command_encode_and_frame(buf, sizeof(buf), ce, args);
-
-    for(int i=0; i<msglen; i++) {
-        r_uart_putc(buf[i]);
-    }
-}
-
 void restore_data(void)
 {
     extern char __data_start, __data_end, __copy_start;
diff --git a/src/ar100/serial.c b/src/ar100/serial.c
index 6fcafdbd8..2c2d89311 100644
--- a/src/ar100/serial.c
+++ b/src/ar100/serial.c
@@ -40,14 +40,10 @@ uint32_t r_uart_fifo_rcv(void){
   return read_reg(R_UART_RFL);
 }
 
-void r_uart_putc(char c){
-  while(!(read_reg(R_UART_LSR) & 1<<5))
-    ;
-  write_reg(R_UART_THR, c);
+uint32_t r_uart_fifo_cantx(void){
+  return read_reg(R_UART_USR) & 1<<1;
 }
 
-void r_uart_puts(char *s){
-  while(*s){
-    r_uart_putc(*s++);
-  }
+void r_uart_putc(char c){
+  write_reg(R_UART_THR, c);
 }
diff --git a/src/ar100/serial.h b/src/ar100/serial.h
index 89eca29e7..2ab0e1e36 100644
--- a/src/ar100/serial.h
+++ b/src/ar100/serial.h
@@ -45,6 +45,4 @@ void r_uart_init(void);
 void r_uart_putc(char c);
 char r_uart_getc(void);
 uint32_t r_uart_fifo_rcv(void);
-void uart_putc(char c);
-void uart_puts(char *s);
-void uart_puti(uint32_t u);
+uint32_t r_uart_fifo_cantx(void);
-- 
2.47.3

